client/components/data/query-active-promotions/index.jsx

```diff
@@ -1,45 +1,25 @@
-/** @format */
-
 /**
  * External dependencies
  */
-
-import PropTypes from 'prop-types';
-import { Component } from 'react';
-import { connect } from 'react-redux';
+import { useEffect, useRef } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
 
 /**
  * Internal dependencies
  */
 import { isRequestingActivePromotions } from 'state/active-promotions/selectors';
 import { requestActivePromotions } from 'state/active-promotions/actions';
 
-class QueryActivePromotions extends Component {
-	UNSAFE_componentWillMount() {
-		if ( ! this.props.requestingActivePromotions ) {
-			this.props.requestActivePromotions();
+export default function QueryActivePromotions() {
+	const requestingActivePromotions = useRef( useSelector( isRequestingActivePromotions ) );
+	const dispatch = useDispatch();
+
+	// Only runs on mount.
+	useEffect( () => {
+		if ( ! requestingActivePromotions.current ) {
+			dispatch( requestActivePromotions() );
 		}
-	}
+	}, [ dispatch ] );
```

Comment:
Can you avoid the ref and do this?
```js
const requesting = useSelector( isRequesting );

useEffect( () => {
  if ( ! requesting ) {
    dispatch( request() );
  }
}, [ dispatch ] );
```
ESLint will probably complain about missing `requesting` variable in the hooks dependencies, but in this case it's wrong.

Even better approach would be moving the `isRequesting` check into the `request` thunk:
```js
const request = () => ( dispatch, getState ) => {
  if ( ! isRequesting( getState() ) {
    dispatch( doRequest() );
  }
}
```

Why? Imagine mounting two query components next to each other:
```js
<QueryActivePromotions />
<QueryActivePromotions />
```
It's unlikely to happen in this form, but in a more indirect form, Calypso does this at a few places like Stats.

This can lead to two requests being issued, because the journey from `useEffect` dispatching the action, through reducer updating the flag and listeners being fired to the query component being rerendered with the new value, that journey can be very complicated. And the `isRequesting` update doesn't rerender the other component soon enough.

With the checks and updates happening outside React lifecycles and effects, purely inside Redux, things are more straightforward and 100% synchronous.

Reply:
Ooooohhh! Now I *finally* understand thunks in Redux and where all this was coming from. That a much cleaner solution, which should work beautifully. Thank you for the thorough explanation, @jsnajdr! 👍